server/src-lib/Hasura/Events/Lib.hs

```diff
@@ -166,49 +168,86 @@ retryAfterHeader = "Retry-After"
-  (,) <$> forkImmortal "pushEvents" logger pushEvents
-      <*> forkImmortal "consumeEvents" logger consumeEvents
+  -> IO void
+processEventQueue logger logenv httpMgr pool getSchemaCache EventEngineCtx{..} = do
+  events0 <- popEventsBatch
+  go events0 0 False
   where
-    -- FIXME proper backpressure. See: #3839 
-    pushEvents = forever $ do
-      let EventEngineCtx q _ _ fetchI = eectx
-      eventsOrError <- runExceptT $ Q.runTx pool (Q.RepeatableRead, Just Q.ReadWrite) fetchEvents
-      case eventsOrError of
-        Left err     -> L.unLogger logger $ EventInternalErr err
-        Right events -> atomically $ mapM_ (TQ.writeTQueue q) events
-      sleep fetchI
-
-    -- TODO this has all events race. How do we know this is correct? Document.
-    consumeEvents = forever $
-      -- ensure async exceptions from link only raised between iterations of forever block:
-      mask_ $ do
-        event <- atomically $ do
-          let EventEngineCtx q _ _ _ = eectx
-          TQ.readTQueue q
-        -- FIXME proper backpressure. See: #3839 
-        t <- async $ runReaderT (processEvent event) (logger, httpMgr, eectx)
-        -- Make sure any stray exceptions are at least logged via 'forkImmortal':
-        link t
-
-    -- NOTE: Blocks in tryWebhook if >= _eeCtxMaxEventThreads invocations active.
+    fetchBatchSize = 100
+    popEventsBatch = do
+      let run = runExceptT . Q.runTx pool (Q.RepeatableRead, Just Q.ReadWrite)
+      run (fetchEvents fetchBatchSize) >>= \case
+          Left err -> do 
+            L.unLogger logger $ EventInternalErr err
+            return []
+          Right events -> 
+            return events
+
+    go :: [Event] -> Int -> Bool -> IO void
+    go events !fullFetchCnt !alreadyWarned = do
+      -- process events ASAP until we've caught up; then we can sleep
+      when (null events) $ sleep _eeCtxFetchInterval
+
+      -- Prefetch next events payload while concurrently working through our current batch.
+      -- NOTE: we probably don't need to prefetch so early, but probably not
+      -- worth the effort for something more fine-tuned
+      eventsNext <- withAsync popEventsBatch $ \eventsNextA -> do
+        -- process approximately in order, minding HASURA_GRAPHQL_EVENTS_HTTP_POOL_SIZE:
+        forM_ events $ \event -> 
+          mask_ $ do
+            atomically $ do  -- block until < HASURA_GRAPHQL_EVENTS_HTTP_POOL_SIZE threads:
+              capacity <- readTVar _eeCtxEventThreadsCapacity
+              check $ capacity > 0
+              writeTVar _eeCtxEventThreadsCapacity (capacity - 1) 
+            let restoreCapacity = liftIO $ atomically $ 
+                  modifyTVar' _eeCtxEventThreadsCapacity (+ 1)
+            t <- async $ flip runReaderT (logger, httpMgr) $ 
+                    processEvent event `finally` restoreCapacity
+            link t
+
+        -- return when next batch ready; some 'processEvent' threads may be running.
+        wait eventsNextA
```

Comment:
IIUC, this call blocks till all events for a batch are finished executing? How would it effect with a slow webhook? Suppose 99 events finished and one event was stuck for 60 seconds (or till timeout) because webhook was very slow. Will the whole queue get blocked for 60 seconds?